src/base/utils/misc.cpp

```diff
@@ -315,30 +317,60 @@ QString Utils::Misc::pythonVersionComplete() {
     return version;
 }
 
-// return best userfriendly storage unit (B, KiB, MiB, GiB, TiB)
+QString Utils::Misc::unitString(Utils::Misc::SizeUnit unit)
+{
+    return QCoreApplication::translate("misc",
+         units[static_cast<int>(unit)].source, units[static_cast<int>(unit)].comment);
+}
+
+// return best userfriendly storage unit (B, KiB, MiB, GiB, TiB, ...)
 // use Binary prefix standards from IEC 60027-2
 // see http://en.wikipedia.org/wiki/Kilobyte
 // value must be given in bytes
 // to send numbers instead of strings with suffixes
-QString Utils::Misc::friendlyUnit(qreal val, bool is_speed)
+bool Utils::Misc::friendlyUnit(qint64 sizeInBytes, qreal &val, Utils::Misc::SizeUnit &unit)
 {
-    if (val < 0)
-        return QCoreApplication::translate("misc", "Unknown", "Unknown (size)");
+    if (sizeInBytes < 0) {
+        return false;
+    }
+
     int i = 0;
-    while(val >= 1024. && i < 4) {
-        val /= 1024.;
+    qreal rawVal = static_cast<qreal>(sizeInBytes);
+
+    while ((rawVal >= 1024.) && (i <= static_cast<int>(SizeUnit::ExbiByte))) {
+        rawVal /= 1024.;
         ++i;
     }
+    val = rawVal;
+    unit = static_cast<SizeUnit>(i);
+    return true;
+}
+
+QString Utils::Misc::friendlyUnit(qint64 bytesValue, bool is_speed)
+{
+    SizeUnit unit;
+    qreal friendlyVal;
+    if (!friendlyUnit(bytesValue, friendlyVal, unit)) {
+        return QCoreApplication::translate("misc", "Unknown", "Unknown (size)");
+    }
     QString ret;
-    if (i == 0)
-        ret = QString::number((long)val) + " " + QCoreApplication::translate("misc", units[0].source, units[0].comment);
+    if (unit == SizeUnit::Byte)
+        ret = QString::number(bytesValue) + " " + unitString(unit);
     else
-        ret = Utils::String::fromDouble(val, 1) + " " + QCoreApplication::translate("misc", units[i].source, units[i].comment);
+        ret = Utils::String::fromDouble(friendlyVal, 1) + " " + unitString(unit);
     if (is_speed)
         ret += QCoreApplication::translate("misc", "/s", "per second");
     return ret;
 }
 
+qlonglong Utils::Misc::sizeInBytes(qreal size, Utils::Misc::SizeUnit unit)
+{
+    for (int i = 0; i < static_cast<int>(unit); ++i) {
```

Comment:
Well, I doubt this will be much simpler. Sure, in one way we can do like
```C++
  return size * (1ll << static_cast<int>(unit)*10);
```
But in the reverse way we have to find the most significant bit of a float number, and I don't know compiler- and platform-independent way to do this. Do you? Otherwise, I believe, there is no much sense to expel this simple loop.